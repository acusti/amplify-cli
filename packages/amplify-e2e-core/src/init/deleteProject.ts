/* eslint-disable import/no-cycle */
import {
  nspawn as spawn, retry, getCLIPath, describeCloudFormationStack,
} from '..';
import { getBackendAmplifyMeta } from '../utils';

/**
 * Runs `amplify delete`
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const deleteProject = async (cwd: string, profileConfig?: any, usingLatestCodebase = false): Promise<void> => {
  // Read the meta from backend otherwise it could fail on non-pushed, just initialized projects
  const { StackName: stackName, Region: region } = getBackendAmplifyMeta(cwd).providers.awscloudformation;
  await retry(
    () => describeCloudFormationStack(stackName, region, profileConfig),
    stack => stack.StackStatus.endsWith('_COMPLETE') || stack.StackStatus.endsWith('_FAILED'),
  );

  const noOutputTimeout = 1000 * 60 * 20; // 20 minutes;
  return spawn(getCLIPath(usingLatestCodebase), ['delete'], { cwd, stripColors: true, noOutputTimeout })
    .wait('Are you sure you want to continue?')
    .sendConfirmYes()
    .wait('Project deleted locally.')
    .runAsync();
};

export const amplifyDeleteWithLongerTimeout = (cwd: string, usingLatestCodebase = false): Promise<void> => new Promise((resolve, reject) => {
  const noOutputTimeout = 1000 * 60 * 30; // 30 minutes;
  spawn(getCLIPath(usingLatestCodebase), ['delete'], { cwd, stripColors: true, noOutputTimeout })
    .wait('Are you sure you want to continue?')
    .sendConfirmYes()
    .wait('Project deleted locally.')
    .run((err: Error) => {
      if (!err) {
        resolve();
      } else {
        reject(err);
      }
    });
});
