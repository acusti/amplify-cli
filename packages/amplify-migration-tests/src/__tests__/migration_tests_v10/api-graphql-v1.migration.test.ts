import { addApiWithoutSchema, 
    addFunction, 
    amplifyPull, 
    amplifyPush, 
    amplifyPushAuth, 
    amplifyPushWithoutCodegen, 
    createNewProjectDir, 
    deleteProject, 
    deleteProjectDir, 
    generateRandomShortId, 
    getAppId, 
    updateApiSchema, 
    updateFunction } from "@aws-amplify/amplify-e2e-core";
import { initJSProjectWithProfileV10 } from "../../migration-helpers-v10/init";
import { 
    assertNoParameterChangesBetweenProjects, 
    collectCloudformationDiffBetweenProjects, 
} from "../../migration-helpers/utils";
import { cfnDiffExclusions } from "../../migration-helpers-v10/cfn-diff-exclusions";

describe('api graphql v1 migration tests', () => {
    let projRoot: string;
    let projectName: string;

    afterEach(async () => {
        await deleteProject(projRoot, undefined, true);
        deleteProjectDir(projRoot);
    });

    // inspired by function_5.test.ts
    it('...adds graphql with v1 transformer, updates models, adds function access, and pulls with latest version', async () => {
        projectName = 'graphqlMigration';
        projRoot = await createNewProjectDir(projectName);
        
        await initJSProjectWithProfileV10(projRoot, { name: projectName, disableAmplifyAppCreation: false });
        await addApiWithoutSchema(projRoot, { transformerVersion: 1 });
        await updateApiSchema(projRoot, projectName, 'simple_model.graphql');
        const fnName = `integtestfn${generateRandomShortId()}`;
        await addFunction(
            projRoot,
            {
                name: fnName,
                functionTemplate: 'Hello World',
            },
            'nodejs',
        );
        await amplifyPush(projRoot);
        await updateApiSchema(projRoot, projectName, 'two-model-schema.graphql');
        await updateFunction(
            projRoot,
            {
                name: fnName,
                functionTemplate: 'Hello World',
                additionalPermissions: {
                permissions: ['storage'],
                choices: ['api', 'function', 'storage'],
                resources: ['Comment:@model(appsync)'],
                resourceChoices: ['Post:@model(appsync)', 'Comment:@model(appsync)'],
                operations: ['read'],
                },
            },
            'nodejs',
        );
        await amplifyPushWithoutCodegen(projRoot, undefined, true);
        
        // pull down with vlatest
        const appId = getAppId(projRoot);
        expect(appId).toBeDefined();
        const projRoot2 = await createNewProjectDir(`${projectName}2`);
        try {
            await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
            // this fails due to a line:
            // "CreateAPIKey": 1,
            // missing in the newly pulled down project's parameters file
            // assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
            expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2, cfnDiffExclusions)).toMatchSnapshot();
            await amplifyPushAuth(projRoot2, true);
            // assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
            expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2, cfnDiffExclusions)).toMatchSnapshot();
        } finally {
            deleteProjectDir(projRoot2);
        }
    });
})