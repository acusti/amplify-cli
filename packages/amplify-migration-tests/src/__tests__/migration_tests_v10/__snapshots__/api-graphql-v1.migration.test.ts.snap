// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api graphql v1 migration tests ...adds graphql with v1 transformer, updates models, adds function access, and pulls with latest version 1`] = `""`;

exports[`api graphql v1 migration tests ...adds graphql with v1 transformer, updates models, adds function access, and pulls with latest version 2`] = `
"Outputs
[+] Output LambdaExecutionRoleArn: {\\"Value\\":{\\"Fn::GetAtt\\":[\\"LambdaExecutionRole\\",\\"Arn\\"]}}

"
`;
